File: server/app/routers/categories.py
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy import select
from sqlalchemy.orm import Session

from app.database import get_db
from app.dependencies.auth import get_current_user, require_admin
from app.models import Category, User
from app.schemas.category import CategoryCreate, CategoryOut, CategoryUpdate

router = APIRouter(prefix="/api/categories", tags=["categories"])


@router.get("", response_model=list[CategoryOut])
def list_categories(
    include_inactive: bool = False,
    db: Session = Depends(get_db),
    user: User = Depends(get_current_user),
):
    query = select(Category).order_by(Category.name)
    if not (include_inactive and user.is_staff):
        query = query.where(Category.is_active.is_(True))
    return db.scalars(query).all()


@router.post("", response_model=CategoryOut, status_code=status.HTTP_201_CREATED)
def create_category(
    payload: CategoryCreate, db: Session = Depends(get_db), _: User = Depends(require_admin)
):
    if db.scalar(select(Category).where(Category.code == payload.code)):
        raise HTTPException(
            status_code=status.HTTP_409_CONFLICT, detail="Category code already exists"
        )
    category = Category(**payload.model_dump())
    db.add(category)
    db.commit()
    db.refresh(category)
    return category


@router.patch("/{category_id}", response_model=CategoryOut)
def update_category(
    category_id: int,
    payload: CategoryUpdate,
    db: Session = Depends(get_db),
    _: User = Depends(require_admin),
):
    category = db.get(Category, category_id)
    if category is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Category not found")
    for key, value in payload.model_dump(exclude_unset=True).items():
        setattr(category, key, value)
    db.commit()
    db.refresh(category)
    return category

File: client/src/services/categories.ts
import api from './api'
import type { Category } from '../types/ticket'

export interface CategoryInput {
  code?: string
  name?: string
  description?: string | null
  is_active?: boolean
}

export async function listCategories(includeInactive = false) {
  const { data } = await api.get<Category[]>('/categories', {
    params: includeInactive ? { include_inactive: true } : undefined,
  })
  return data
}

export async function createCategory(input: CategoryInput) {
  const { data } = await api.post<Category>('/categories', input)
  return data
}

export async function updateCategory(id: number, input: CategoryInput) {
  const { data } = await api.patch<Category>(`/categories/${id}`, input)
  return data
}
